### ISSUE TITLE
Memory Usage

### ISSUE THREAD

**Issue by Caseyc585** [canton7/RestEase#74]
This might be more of a question than an issue... anyways:

I'm using RestEase as part of a DotNetCore 2.0 WebAPI to reach out to other rest apis. I love the simplicity but I notice that my API has a memory leak and I traced it back to my usage of Restease.

```
public class DemoController : Controller
    {
        public ILegacyAPI legacyAPI { get; set; }

        public DemoController(IOptions<AppSettingsServices> options)
        {
            this.legacyAPI = RestClient.For<ILegacyAPI>(options.Value.LegacyAPIBaseUrl);
        }

        public async Task<dynamic> CallOutToOtherAPI()
        {
            var response = await legacyAPI.GetSomethingFromOtherAPI();

            //Do stuff with response and return
        }
    }
```

Should I be using a scoped rest client maybe? I also saw something about implementing IDisposable?

**Comment by fazouane-marouane** [comment]
Hi

The issue here is that you're creating a new rest client at each call to your webapi.
What you can do is treat your ILegacyAPI as a service contract and inject it into your controller like you'd do for other services. It's highly recommended to declare it as a singleton service.

For `IDisposable`, you can say that your `ILegacyAPI` extends `IDisposable` then you can manually dispose of your rest client. That'll solve your memory leak but it'll introduce performance issues since you'll be creating-disposing. You can view the frequent creation-dispose as an antipattern. You should use in your case the dependency injection mechanisms of aspcore like explained.

**Comment by canton7** [comment]
I wouldn't have expected that to leak memory. It's true that people recommend keeping a HttpClient instance around, rather than instantiating a new one every time you need one -- you can do that either by keeping the ILegacyApi instance around, or by keeping a static HttpClient and passing it to the `RestClient.For` method.

**Comment by Caseyc585** [comment]
Here's what I tried for aspnetcore in the startup:

```
services.AddSingleton(RestClient.For<ILegacyAPI>("base url here"));
```

And then in the controller:
```
public class DemoController : Controller
    {
        private readonly ILegacyAPI legacyAPI;

        public DemoController(ILegacyAPI legacyAPI)
        {
            this.legacyAPI = legacyAPI;
        }

        public async Task<dynamic> CallOutToOtherAPI()
        {
            var response = await legacyAPI.GetSomethingFromOtherAPI();

            //Do stuff with response and return
        }
    }
```

Is that right?  I'm still seeing memory issues after time. I did also try to throw together a quick console app to validate what you guys are saying about keeping the client around as long as you're using it.

```
static void Main(string[] args)
        {
            var task = Task.Factory.StartNew(async () =>
            {
                var client = RestEase.RestClient.For<ILegacyAPI>("<URL Here>");
                while (true)
                {
                    try
                    {
                        var response = await client.GetData();
                        System.Threading.Thread.Sleep(500);
                    }
                    catch(Exception e)
                    {
                        var catchstr = e.ToString();
                    }                    
                }
            });

            Console.ReadKey();
        }
```


I do see stable memory usage in that example.  So I wonder if my singleton usage isn't correct?

**Comment by canton7** [comment]
What memory issues exactly are you seeing, and what makes you think that RestEase is responsible?

The other thing I notice is that your controller calls `client.GetSomethingFromOtherAPI()`, while your demo calls `client.GetData()`. Is the return type different between those methods?

**Comment by fazouane-marouane** [comment]
Oh right. If you’re also getting a raw http response, you should dispose of it after getting what you wanted from it. (@canton7 is this what you were thinking of?)

**Comment by canton7** [comment]
Perhaps, perhaps it's an `Response<T>` (same issue). Perhaps it's a custom type which has something funky about it.

**Comment by Caseyc585** [comment]
@canton7 - the calls are different cause I gave you example code. 

What I've noticed is that I started having memory issues when I started using Restease.  I was able prove out with a console app that they way I was using restease was causing memory issues. 

I'm seeing its difficult to pin down what is actually causing the issue though.  Here's what I'm seeing in the memory profiler from that console app:

![screen shot 2017-11-07 at 8 34 26 am](https://user-images.githubusercontent.com/4412526/32496256-9f49b9d8-c396-11e7-89f4-b002bf224ce8.png)

Here's the code that I was using for that snap shot:

```
class Program
    {
        static void Main(string[] args)
        {
            var task = Task.Factory.StartNew(async () =>
            {
                while (true)
                {
                    var client = RestEase.RestClient.For<IAPIClient>("<base url here>");
                    try
                    {
                        var response = await client.GetData();
                        System.Threading.Thread.Sleep(500);
                    }
                    catch(Exception e)
                    {
                        var catchstr = e.ToString();
                    }                    
                }
            });

            Console.ReadKey();
        }
    }
```

**Comment by canton7** [comment]
> @canton7 - the calls are different cause I gave you example code.

Are the return types actually different, or are they the same? I want to make sure that you are using RestEase in the same way both in your controller and in your test code.

> I was able prove out with a console app that they way I was using restease was causing memory issues.

I'm afraid I'm still unclear on where there's a problem. Previously you could not reproduce an issue in the test console app. Now it looks like you have been able to reproduce them?

There's nothing scary in the profile dump you posted that I can see -- you're just looking at [these static cached MethodInfo/PropertyInfo/ConstructorInfo fields](https://github.com/canton7/RestEase/blob/master/src/RestEase/Implementation/ImplementationBuilder.cs#L31).

**Comment by fazouane-marouane** [comment]
The example reproduces, naturally, because you create a client at every iteration.

**Comment by canton7** [comment]
Hi, any response?

**Comment by Caseyc585** [comment]
I'm still seeing some memory consumption, but I wonder if its delayed release or what. Would this be the right way to do it for asp.net core?

`services.AddSingleton(RestClient.For<ILegacyAPI>("base url here"));`

or should I make my HttpClient the singleton?

**Comment by fazouane-marouane** [comment]
That’s the correct way. But you still need to show us the exact interface you are using. If you expect a Response<T> for example, then you need to clean up the response object after each call.

**Comment by canton7** [comment]
I'm still interested in what you have seen which makes you think that there's a memory leak. The screen shot you posted above only shows a snail number of cached Type objects.

**Comment by MaxDeg** [comment]
Hello,

We also face an issue with memory usage. It seems it come from the unmanaged heap :(
I attached to this ticket a test project reproducing the issue and on the following screenshot you could see the "private bytes" counter that constantly increase during the run of this test.

Hoping this could help.

![image](https://user-images.githubusercontent.com/6094373/37199591-537b3b52-2382-11e8-9586-08b37d44bf99.png)

[RestEaseMemoryLeakTest.zip](https://github.com/canton7/RestEase/files/1796214/RestEaseMemoryLeakTest.zip)

**Comment by MaxDeg** [comment]
A colleague of mine found the issue :) 
https://github.com/canton7/RestEase/blob/master/src/RestEase/Implementation/ImplementationBuilder.cs#L83

This line is creating a new instance each time we are accessing the Instance property. With this fix it should be much better. 
```csharp
public static ImplementationBuilder Instance {get; } = new ImplementationBuilder();
```

**Comment by canton7** [comment]
Yes, apparently I'm an idiot (at least on a part-time basis). I'll get that fixed up over lunch.

**Comment by MaxDeg** [comment]
I'm working on the patch. I can create the PR if you want.
This is an error prone addition of C# 7, it's very easy to get tricked by it.

**Comment by canton7** [comment]
Relesed as 1.4.5.
